File: frontend/src/utils/permissions.js
export const ADMIN_ROLES = ['superadmin', 'admin']
export const MANAGER_ROLES = ['superadmin', 'admin', 'manager']
export const STAFF_ROLES = ['superadmin', 'admin', 'manager', 'lead', 'mentor']

// Admin is view-only (can see everything but cannot create/edit/delete except staff registration)
export const isAdminViewOnly = (role) => role === 'admin'
export const canWrite = (role) => role !== 'admin'  // everyone except admin can write

export const canManageUsers = (role) => MANAGER_ROLES.includes(role)
export const canAddStaff = (role) => ADMIN_ROLES.includes(role)
export const canViewPayroll = (role) => MANAGER_ROLES.includes(role)
export const canCreateTask = (role) => canWrite(role) && ['mentor', 'lead', ...MANAGER_ROLES].includes(role)
export const canVerifyTask = (role) => canWrite(role) && ['mentor', 'lead', ...MANAGER_ROLES].includes(role)
export const canApproveDocs = (role) => canWrite(role) && ['lead', ...MANAGER_ROLES].includes(role)
export const canApproveLeave = (role) => canWrite(role) && ['mentor', 'lead', ...MANAGER_ROLES].includes(role)
export const canManageOrg = (role) => ADMIN_ROLES.includes(role)
export const isIntern = (role) => role === 'intern'
export const isStaffOrAbove = (role) => STAFF_ROLES.includes(role)

File: backend/apps/permissions.py
from rest_framework.permissions import BasePermission

ADMIN_ROLES = ('superadmin', 'admin')
MANAGER_ROLES = ('superadmin', 'admin', 'manager')
STAFF_ROLES = ('superadmin', 'admin', 'manager', 'lead', 'mentor')

# Roles that can perform write operations (admin excluded — view-only)
WRITE_ROLES = ('superadmin', 'manager', 'lead', 'mentor')


class IsAdminOrManager(BasePermission):
    def has_permission(self, request, view):
        if not request.user.is_authenticated:
            return False
        profile = getattr(request.user, 'profile', None)
        return request.user.is_staff or (profile and profile.role in MANAGER_ROLES)


class IsAdminOnly(BasePermission):
    def has_permission(self, request, view):
        if not request.user.is_authenticated:
            return False
        profile = getattr(request.user, 'profile', None)
        return request.user.is_superuser or (profile and profile.role in ADMIN_ROLES)


class IsStaffOrAbove(BasePermission):
    """Lead, Mentor, Manager, Admin"""
    def has_permission(self, request, view):
        if not request.user.is_authenticated:
            return False
        profile = getattr(request.user, 'profile', None)
        return profile and profile.role in STAFF_ROLES


class IsMentor(BasePermission):
    def has_permission(self, request, view):
        if not request.user.is_authenticated:
            return False
        profile = getattr(request.user, 'profile', None)
        return profile and profile.role in STAFF_ROLES


class IsOwnerOrAdminOrManager(BasePermission):
    def has_object_permission(self, request, view, obj):
        if not request.user.is_authenticated:
            return False
        profile = getattr(request.user, 'profile', None)
        if profile and profile.role in MANAGER_ROLES:
            return True
        user_field = getattr(obj, 'user', None) or getattr(obj, 'employee', None) or getattr(obj, 'intern', None)
        return user_field == request.user


class ReadOnly(BasePermission):
    def has_permission(self, request, view):
        return request.method in ('GET', 'HEAD', 'OPTIONS')


class DenyAdminWrite(BasePermission):
    """
    Admin role gets view-only access.
    Returns False for write methods (POST/PUT/PATCH/DELETE) when user is admin.
    superadmin is NOT restricted.
    """
    def has_permission(self, request, view):
        if request.method in ('GET', 'HEAD', 'OPTIONS'):
            return True
        if not request.user.is_authenticated:
            return True  # Let other permission classes handle auth
        profile = getattr(request.user, 'profile', None)
        if profile and profile.role == 'admin':
            return False  # Block admin write
        return True  # Allow everyone else


class IsAdminOrManagerOrStaff(BasePermission):
    def has_permission(self, request, view):
        if not request.user.is_authenticated:
            return False
        profile = getattr(request.user, 'profile', None)
        return request.user.is_superuser or (profile and profile.role in STAFF_ROLES)

